device/mock_device/src/mock_device_ffi.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 Comcast Cable Communications Management, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use jsonrpsee::core::server::rpc_module::Methods;
+use ripple_sdk::{
+    api::status_update::ExtnStatus,
+    crossbeam::channel::Receiver as CReceiver,
+    export_channel_builder, export_extn_metadata, export_jsonrpc_extn_builder,
+    extn::{
+        client::{extn_client::ExtnClient, extn_sender::ExtnSender},
+        extn_id::{ExtnClassId, ExtnId},
+        ffi::{
+            ffi_channel::{ExtnChannel, ExtnChannelBuilder},
+            ffi_jsonrpsee::JsonRpseeExtnBuilder,
+            ffi_library::{CExtnMetadata, ExtnMetadata, ExtnSymbolMetadata},
+            ffi_message::CExtnMessage,
+        },
+    },
+    framework::ripple_contract::{ContractFulfiller, RippleContract},
+    log::{debug, error, info},
+    semver::Version,
+    tokio::{self, runtime::Runtime, sync::Mutex},
+    utils::{error::RippleError, logger::init_logger},
+};
+
+use crate::{
+    mock_device_controller::{MockDeviceController, MockDeviceControllerServer},
+    mock_device_ws_server_processor::MockDeviceMockWebsocketServerProcessor,
+    utils::{boot_ws_server, load_mock_data},
+};
+
+const EXTN_NAME: &str = "mock_device";
+
+fn init_library() -> CExtnMetadata {
+    let _ = init_logger(EXTN_NAME.into());
+
+    let mock_device_channel = ExtnSymbolMetadata::get(
+        ExtnId::new_channel(ExtnClassId::Device, EXTN_NAME.into()),
+        ContractFulfiller::new(vec![RippleContract::MockWebsocketServer]),
```

Comment:
Can a given extension have multiple ws mock server? How can it identify between them?

Reply:
Not sure I understand the question here. I suppose the extension could start multiple websocket servers if needed but its not setup to at the moment. They would all try listening on the gateway in the Platform Parameters anyway so would fail to start.